nexus/src/app/sagas/demo.rs

```diff
@@ -0,0 +1,126 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Demo saga, used for testing and interactive debugging
+//!
+//! The "Demo" saga exists so that developers and automated tests can create a
+//! saga that will not complete until they take some action to complete it.  The
+//! saga just waits until it gets the message that it should finish.  Users
+//! create demo sagas and complete them using requests to the internal API.
+//!
+//! The implementation is entirely in-memory, which means you have to send the
+//! completion message to the Nexus that's running the saga.  However, it does
+//! work across Nexus restarts, so this can be used to exercise the saga
+//! recovery path.
+//!
+//! It's tempting to build this only for development and not official releases,
+//! but that'd be more work, there's little downside to always including it, and
+//! it's conceivable that it'd be useful for production systems, too.
+
+use super::NexusActionContext;
+use super::{ActionRegistry, NexusSaga, SagaInitError};
+use crate::app::sagas::declare_saga_actions;
+use omicron_common::api::external::Error;
+use omicron_uuid_kinds::DemoSagaUuid;
+use serde::Deserialize;
+use serde::Serialize;
+use slog::info;
+use std::collections::BTreeMap;
+use steno::ActionError;
+use tokio::sync::oneshot;
+
+/// Set of demo sagas that have been marked completed
+///
+/// Nexus maintains one of these at the top level.  Individual demo sagas wait
+/// until their id shows up here, then remove it and proceed.
+pub struct CompletingDemoSagas {
+    ids: BTreeMap<DemoSagaUuid, oneshot::Sender<()>>,
+}
+
+impl CompletingDemoSagas {
+    pub fn new() -> CompletingDemoSagas {
+        CompletingDemoSagas { ids: BTreeMap::new() }
+    }
+
+    pub fn complete(&mut self, id: DemoSagaUuid) -> Result<(), Error> {
+        self.ids
+            .remove(&id)
+            .ok_or_else(|| {
+                Error::non_resourcetype_not_found(format!(
+                    "demo saga with id {:?}",
+                    id
+                ))
+            })?
+            .send(())
+            .map_err(|_| {
+                Error::internal_error(
+                    "saga stopped listening (Nexus shutting down?)",
+                )
+            })
+    }
+
+    pub fn subscribe(&mut self, id: DemoSagaUuid) -> oneshot::Receiver<()> {
+        let (tx, rx) = oneshot::channel();
+        assert!(
```

Comment:
hmm, I kind of feel like maybe this should return an error that results in the demo saga unwinding, rather than panicking? I agree that the panic _shouldn't happen_ ever, as far as I can tell, but it also seems a little bit unfortunate for the test endpoint to be able to (theoretically) crash a nexus...I dunno if it matters.

Reply:
I'm ambivalent about trying to handle programmer errors programmatically but I went ahead and made this change.